docs/content/guides/developer/app-examples/plinko.mdx

```diff
@@ -0,0 +1,772 @@
+---
+title: Plinko
+---
+
+Plinko is an example implementation of the popular casino game of the same name on Sui. This guide walks through its components, providing a detailed look at the module's functions, structures, constants, and their significance in the overall gameplay mechanism.
+
+Building an on-chain Plinko game shares a lot of similarities with the [Coin Flip](./coin-flip.mdx) game and [Blackjack](./blackjack.mdx) game. For that reason, this example covers only the smart contracts (Move modules) and frontend logic.
+
+## Gameplay
+
+In the Plinko game, players start by choosing how many Plinko balls to play and setting their stake for each one. This decision strategically influences the potential payout and risk level. When the game starts, blockchain technology ensures each ball's path is randomly determined as it descends through a pegged board. The objective is to guide the balls into slots at the board's base, each associated with different reward multipliers. The smart contract automatically computes the game's outcome, which evaluates the final positions of the balls to calculate the player's total earnings. This process involves cryptographic techniques, such as BLS signatures for randomness and verifiable random function (VRF) inputs for unique game rounds, ensuring transparency and integrity in the game's execution.
+
+## Backend
+
+Follow the comments in each module's code to understand the logic each creates.
+
+### plinko::plinko
+
+The `plinko::plinko` module combines various Sui blockchain features, such as coin handling, event emissions, and dynamic object fields, to create on-chain Plinko games.
+
+```move title="plinko.move"
+module plinko::plinko {
```

Comment:
**Updated Contract Version:**

// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module plinko::plinko {
    // === Imports ===
    use std::vector;
    use sui::coin::{Self, Coin};
    use sui::balance::{Self, Balance};
    use sui::sui::SUI;
    use sui::bls12381::bls12381_min_pk_verify;
    use sui::object::{Self, UID, ID};
    use sui::tx_context::{Self, TxContext};
    use sui::transfer;
    use sui::event::emit;
    use sui::hash::{blake2b256};
    use sui::dynamic_object_field::{Self as dof};

    // Import Counter NFT module
    use plinko::counter_nft::{Self, Counter};

    // Import HouseData module
    use plinko::house_data::{Self as hd, HouseData};

    // === Errors ===
    const EStakeTooLow: u64 = 0;
    const EStakeTooHigh: u64 = 1;
    const EInvalidBlsSig: u64 = 2;
    const EInsufficientHouseBalance: u64 = 5;
    const EGameDoesNotExist: u64 = 6;

    // === Structs ===

    /// Represents a game and holds the acrued stake.
    struct Game has key, store {
        id: UID,
        game_start_epoch: u64,
        stake: Balance<SUI>,
        player: address,
        // The VRF input used to generate the extended beacon
        vrf_input: vector<u8>,
        fee_bp: u16
    }

    // === Events ===

    /// Emitted when a new game has started.
    struct NewGame has copy, drop {
        game_id: ID,
        player: address,
        vrf_input: vector<u8>,
        user_stake: u64,
        fee_bp: u16
    }

    /// Emitted when a game has finished.
    struct Outcome has copy, drop {
        game_id: ID,
        result: u64,
        player: address,
        // The trace path of the extended beacon
        trace: vector<u8> 
    }

    // === Public Functions ===

    /// Function used to create a new game. The player must provide a Counter NFT and the number of balls.
    public fun start_game(counter: &mut Counter, num_balls: u64, coin: Coin<SUI>, house_data: &mut HouseData, ctx: &mut TxContext): ID {
        let fee_bp = hd::base_fee_in_bp(house_data);
        let (id, new_game) = internal_start_game(counter, num_balls, coin, house_data, fee_bp, ctx);
        dof::add(hd::borrow_mut(house_data), id, new_game);
        id
    }

    /// Completes the game by calculating the outcome and transferring the funds to the player.
    /// The player must provide a BLS signature of the VRF input and the number of balls to calculate the outcome.
    /// It emits an Outcome event with the game result and the trace path of the extended beacon.
    public fun finish_game(game_id: ID, bls_sig: vector<u8>, house_data: &mut HouseData, num_balls: u64, ctx: &mut TxContext): (u64, address, vector<u8>) {
        // Ensure that the game exists.
        assert!(game_exists(house_data, game_id), EGameDoesNotExist);

        // Retrieves and removes the game from HouseData, preparing for outcome calculation.
        let Game {
            id,
            game_start_epoch: _,
            stake,
            player,
            vrf_input,
            fee_bp: _
        } = dof::remove<ID, Game>(hd::borrow_mut(house_data), game_id);

        object::delete(id);

        // Validates the BLS signature against the VRF input.
        let is_sig_valid = bls12381_min_pk_verify(&bls_sig, &hd::public_key(house_data), &vrf_input);
        assert!(is_sig_valid, EInvalidBlsSig);

        // Initialize the extended beacon vector and a counter for hashing.
        let extended_beacon = vector::empty<u8>();
        let counter: u8 = 0;

        // Extends the beacon until it has enough data for all ball outcomes.
        while (vector::length(&extended_beacon) < (num_balls * 12)) {
            // Create a new vector combining the original BLS signature with the current counter value.
            let hash_input = vector::empty<u8>();
            vector::append(&mut hash_input, bls_sig);
            vector::append(&mut hash_input, vector::singleton(counter));
            // Generate a new hash block from the unique hash input.
            let block = blake2b256(&hash_input);
            // Append the generated hash block to the extended beacon.
            vector::append(&mut extended_beacon, block);
            // Increment the counter for the next iteration to ensure a new unique hash input.
            counter = counter + 1;
        };

        // Initializes variables for calculating game outcome.
        let trace = vector::empty<u8>();
        // Calculate the stake amount per ball
        let stake_per_ball = balance::value<SUI>(&stake) / num_balls;
        let total_funds_amount: u64 = 0;

        // Calculates outcome for each ball based on the extended beacon.
        let ball_index = 0;
        while (ball_index < num_balls) {
            let state: u64 = 0;
            let i = 0;
            while (i < 12) {
                // Calculate the byte index for the current ball and iteration.
                let byte_index = (ball_index * 12) + i;
                // Retrieve the byte from the extended beacon.
                let byte = *vector::borrow(&extended_beacon, byte_index);
                // Add the byte to the trace vector
                vector::push_back<u8>(&mut trace, byte);
                // Count the number of even bytes
                // If even, add 1 to the state
                // Odd byte -> 0, Even byte -> 1
                // The state is used to calculate the multiplier index
                state = if (byte % 2 == 0) { state + 1 } else { state };
                i = i + 1;
            };

        // Calculate multiplier index based on state
        let multiplier_index = state % vector::length(&hd::multiplier(house_data));
        // Retrieve the multiplier from the house data
        let result = *vector::borrow(&hd::multiplier(house_data), multiplier_index);
        
        // Calculate funds amount for this particular ball
        // Divide by 100 to adjust for multiplier scale and SUI units
        let funds_amount_per_ball = (result * stake_per_ball)/100; 
        // Add the funds amount to the total funds amount
        total_funds_amount = total_funds_amount + funds_amount_per_ball;
        ball_index = ball_index + 1;
    };

    // Processes the payout to the player and returns the game outcome.
    let payout_balance_mut = hd::borrow_balance_mut(house_data);
    let payout_coin = coin::take(payout_balance_mut, total_funds_amount, ctx);

    balance::join(payout_balance_mut, stake);

    // transfer the payout coins to the player
    transfer::public_transfer(payout_coin, player);
    // Emit the Outcome event
    emit(Outcome {
        game_id,
        result: total_funds_amount,
        player,
        trace
    });

    // return the total amount to be sent to the player, (and the player address)
    (total_funds_amount, player, trace)
}

    // === Public-View Functions ===

    /// Returns the epoch in which the game started.
    public fun game_start_epoch(game: &Game): u64 {
        game.game_start_epoch
    }

    /// Returns the total stake.
    public fun stake(game: &Game): u64 {
        balance::value(&game.stake)
    }

    /// Returns the player's address.
    public fun player(game: &Game): address {
        game.player
    }

    /// Returns the player's vrf_input bytes.
    public fun vrf_input(game: &Game): vector<u8> {
        game.vrf_input
    }

    /// Returns the fee of the game.
    public fun fee_in_bp(game: &Game): u16 {
        game.fee_bp
    }

    // === Admin Functions ===

    /// Helper function to check if a game exists.
    public fun game_exists(house_data: &HouseData, game_id: ID): bool {
        dof::exists_(hd::borrow(house_data), game_id)
    }

    /// Helper function to check that a game exists and return a reference to the game Object.
    /// Can be used in combination with any accessor to retrieve the desired game field.
    public fun borrow_game(game_id: ID, house_data: &HouseData): &Game {
        assert!(game_exists(house_data, game_id), EGameDoesNotExist);
        dof::borrow(hd::borrow(house_data), game_id)
    }

    // === Private Functions ===

    /// Internal helper function used to create a new game. 
    /// The player must provide a guess and a Counter NFT.
    /// Stake is taken from the player's coin and added to the game's stake. 
    fun internal_start_game(counter: &mut Counter, num_balls: u64, coin: Coin<SUI>, house_data: &HouseData, fee_bp: u16, ctx: &mut TxContext): (ID, Game) {
        let user_stake = coin::value(&coin);
        // Ensure that the stake is not higher than the max stake.
        assert!(user_stake <= hd::max_stake(house_data), EStakeTooHigh);
        // Ensure that the stake is not lower than the min stake.
        assert!(user_stake >= hd::min_stake(house_data), EStakeTooLow);
        // Ensure that the house has enough balance to play for this game.
        assert!(hd::balance(house_data) >= (user_stake*(*vector::borrow(&hd::multiplier(house_data), 0)))/100, EInsufficientHouseBalance);

        // Get the VRF input and increment the counter
        let vrf_input = counter_nft::get_vrf_input_and_increment(counter, num_balls);

        let id = object::new(ctx);
        let game_id = object::uid_to_inner(&id);

        // Create a new game object and emit a NewGame event.
        let new_game = Game {
            id,
            game_start_epoch: tx_context::epoch(ctx),
            stake: coin::into_balance<SUI>(coin),
            player: tx_context::sender(ctx),
            vrf_input,
            fee_bp
        };
        // Emit a NewGame event
        emit(NewGame {
            game_id,
            player: tx_context::sender(ctx),
            vrf_input,
            user_stake,
            fee_bp
        });

        (game_id, new_game)
    }
}